bfdd/ptm_adapter.c

```diff
@@ -128,6 +128,35 @@ static void debug_printbpc(const struct bfd_peer_cfg *bpc, const char *fmt, ...)
 		   timers[0], timers[1], timers[2], cbit_str);
 }
 
+static void _ptm_bfd_session_del(struct bfd_session *bs, uint8_t diag)
+{
+	if (bglobal.debug_peer_event)
+		zlog_debug("session-delete: %s", bs_to_string(bs));
+
+	/* Change state and notify peer. */
+	bs->ses_state = PTM_BFD_DOWN;
+	bs->local_diag = diag;
+	ptm_bfd_snd(bs, 0);
+
+	/* Session reached refcount == 0, lets delete it. */
+	if (bs->refcount == 0) {
+		/*
+		 * Sanity check: if there is a refcount bug, we can't delete
+		 * the session a user configured manually. Lets leave a
+		 * message here so we can catch the bug if it exists.
+		 */
+		if (CHECK_FLAG(bs->flags, BFD_SESS_FLAG_CONFIG)) {
+			zlog_err(
+				"ptm-del-session: [%s] session refcount is "
+				"zero but it was configured by CLI",
+				bs_to_string(bs));
```

Comment:
is this not assert candidate?

Reply:
I think it is better to leave as an error: BFD integration code could improve how it handles the state/message sending, however even then it is possible that we ever make a mistake and send duplicated de registration messages.

I'm not saying this is happening (no reports and I never saw it happen), but it is better to be safe than sorry and I don't see the benefit of crashing the daemon for the user. If we start getting duplicated (or more) de registration message its possible we'll see other problems too.

Reply:
I suppose the problem is a refcount mismatch will either result in a memory leak or referencing unallocated memory, etc., correct? Could such a mismatch cause a crash later on? If so, maybe it's worth crashing the daemon here, when we see the problem, rather than someplace else, then needing to look through the logs to see if this error message pops up?

If the answer is -- "this can't cause a crash anyway," then I think an error is fine here... If the answer is "this might cause a crash someplace else," then I think an assert might be better here ... ??

Reply:
> I suppose the problem is a refcount mismatch will either result in a memory leak or referencing unallocated memory, etc., correct?

No. Worst case is you'll have a peer stuck, but you can always reset all BFD configuration with `no bfd` at the config node or attempt to manually remove it (e.g. `no peer foo` in BFD config node). Then if nothing works you can set it to `shutdown`.

NOTE: I never saw and I don't expect to see `refcount` go more than the peers that asked for it, because the code handles multiple registration from same PID. This code is only for handling refcount going below `0` with CLI in which case it is recoverable via CLI commands.


> Could such a mismatch cause a crash later on?

No. The peer will continue to operate normally, you can always remove it like I said previously.


> then needing to look through the logs to see if this error message pops up?

I want to leave it in logs in case someone ever hits this. If some daemon somehow causes multiple same destination de registration then I want to know about it.

Another NOTE: I didn't change `refcount` code behavior, I just improved it a little bit by putting it in a separated function. I even kept the `ptm-del-session` log message prefix to know it is the same deletion code path as before.